src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -66,12 +76,57 @@ RouterExecutorStart(QueryDesc *queryDesc, int eflags, Task *task)
 	/* ensure that the task is not NULL */
 	Assert(task != NULL);
 
-	/* disallow transactions and triggers during distributed modify commands */
+	/* disallow triggers during distributed modify commands */
 	if (commandType != CMD_SELECT)
 	{
-		bool topLevel = true;
-		PreventTransactionChain(topLevel, "distributed commands");
 		eflags |= EXEC_FLAG_SKIP_TRIGGERS;
+
+		if (IsTransactionBlock())
+		{
+			if (shardForTransaction == INVALID_SHARD_ID)
```

Comment:
Can we move this logic into a separate function, we're 6 blocks indented below...